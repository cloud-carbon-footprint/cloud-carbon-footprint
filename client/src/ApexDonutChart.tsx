import { sumCO2ByService } from './transformData'
import Chart from 'react-apexcharts'
import React, { FunctionComponent } from 'react'
import { EstimationResult } from './types'

export const ApexDonutChart: FunctionComponent<ApexDonutChartProps> = ({ data }) => {
  const donutData = sumCO2ByService(data)

  const options = {
    height: '500px',
    labels: Object.keys(donutData),
    legend: {
      position: 'top',
      offsetY: -8,
    },
    series: Object.values(donutData),
    title: {
      text: 'Carbon Emissions by Service',
      offsetY: -8,
      style: {
        fontSize: '24px',
      },
    },
    width: '100%',
    tooltip: {
      fillSeriesColor: false,
      y: {
        formatter: function (value: string) {
          return `${value} kg CO2e`
        },
      },
    },
  }
  return <Chart options={options} series={options.series} type="donut" width={options.width} />
}

type ApexDonutChartProps = {
  data: EstimationResult[]
}
