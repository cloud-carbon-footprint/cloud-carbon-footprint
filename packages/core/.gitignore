dist
*.csv
*.cache.json
.env
.env.production
