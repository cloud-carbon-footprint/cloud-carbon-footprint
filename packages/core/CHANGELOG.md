# @cloud-carbon-footprint/core

## 0.17.3

### Patch Changes

- 65386330: bumps ramda dependency

## 0.17.2

### Patch Changes

- f1818030: Fixes pointer for main file

## 0.17.1

### Patch Changes

- 9e929f03: Fixes main field in package.json pointing to incorrect file
- Updated dependencies [1d297700]
- Updated dependencies [3f17990d]
  - @cloud-carbon-footprint/common@1.10.0

## 0.17.0

### Minor Changes

- a6423a68: Aggregate footprint estimates for AWS resources by tags.

### Patch Changes

- a6423a68: A small refactor to fix one of the null check errors by replacing a null date.
- a6423a68: updates ts-node dependency
- a6423a68: Refactored to resolve null check errors and make code more efficient.
- a6423a68: Add default values to get rid of null check errors.
- Updated dependencies [a6423a68]
- Updated dependencies [a6423a68]
- Updated dependencies [a6423a68]
  - @cloud-carbon-footprint/common@1.8.0

## 0.16.1

### Patch Changes

- 510d4b86: updates packages: axios googleapis dotenv @types/jest-when

  See [this commit](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/commit/8d8c1db6ff94da5127d559e10632479a8520c67a) for changes to the create-app templates.

- ff05607b: Bumps to latest version of typescript and sets resolution for this

  See [this commit](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/commit/6cdc1469dcd5380c9c8a84b9fe13b977991db54c) for changes to the create-app templates.

- Updated dependencies [510d4b86]
- Updated dependencies [ff05607b]
  - @cloud-carbon-footprint/common@1.5.1

## 0.16.0

### Minor Changes

- 497ae495: Adds support for estimating energy and carbon emissions for Azure GPU VMs
- ef6af294: Adds initial support for energy/carbon estimation for GPU instances for AWS and GCP
- 9938c9b0: refactors ccf for v1 implementation of on-premise estimations
  Refer to [this](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/commit/b3ba4120d633a8b83bf8bc0c131855dd67e6a288) commit to update cli package templates.

### Patch Changes

- Updated dependencies [7ecd432d]
- Updated dependencies [9fcbfc67]
- Updated dependencies [9938c9b0]
  - @cloud-carbon-footprint/common@1.5.0

## 0.15.0

### Minor Changes

- a23e1e59: Updates logic for handling unknown usage types to dynamically build kilowatt hour per usage amount (GCP, Azure) or cost (AWS) by service name and usage unit, to increase accuracy.

### Patch Changes

- Updated dependencies [f40ce29e]
- Updated dependencies [04f2be16]
  - @cloud-carbon-footprint/common@1.4.0

## 0.14.0

### Minor Changes

- 808085cc: Adds replication factor for Cloud Spanner and Kubernetes engine and updates unsoported usage types
- aaeb61a3: Updates estimation for unknown usage types to calculate kilowatt hours first, then co2e, to improve acccuracy

### Patch Changes

- 8fd171ed: Updates a number of packages and fixes linting, typescript and dependency issues
- Updated dependencies [8fd171ed]
  - @cloud-carbon-footprint/common@1.3.1

## 0.13.0

### Minor Changes

- c29a3b53: Adds support for specifying groupBy via API param and for displaying line chart data according to data grouping

  For changes to create-app templates, please refer to this [commit](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/commit/8743e9a36f005716095300b7a1f331b4ffaa8100).

### Patch Changes

- Updated dependencies [c29a3b53]
  - @cloud-carbon-footprint/common@1.3.0

## 0.12.1

### Patch Changes

- a304acb6: Upgrades version of typescript to 4.5.3
- Updated dependencies [a304acb6]
  - @cloud-carbon-footprint/common@1.2.1

## 0.12.0

### Minor Changes

- 4238d3b8: "Add embodied emissions to the estimations for GCP"

### Patch Changes

- 4238d3b8: Bug fix in core package to handle no estimates and implementation of terraform deployments for aws
- 4238d3b8: changeset: Adds embodied emissions to the estimations for AWS and Azure

## 0.11.1

### Patch Changes

- 8b200348: Renames numberOfvCpus in ComputeUsage to match its use in ComputeEstimator

## 0.11.0

### Minor Changes

- c439b0ba: implements reclassification and estimations for unknown usage rows
- 52237bc6: Adds additional dashboard for viewing cloud provider recommendations including refactoring filters for reusability

  For updating the create-app templates, please refer to the following [commit](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/commit/d1f9a94ea88e6f9210a781e16df18b9f64a0b03d).

### Patch Changes

- Updated dependencies [f3569daa]
- Updated dependencies [61332214]
- Updated dependencies [52237bc6]
  - @cloud-carbon-footprint/common@1.2.0

## 0.10.0

### Minor Changes

- 5a0aab5b: implements classifying unknown usage types for aws

## 0.9.0

### Minor Changes

- 13f39ac7: implements recommendations for aws rightsizing recommendations api

  Please refer to this [commit](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/commit/b2181a1942b0910613c8b15f97d074a7be6100b8) to update create app templates.

- e76d5fdd: Adds accountId to the Estimation Results returned from the Cloud Carbon Footprint API

  Please refer to this [commit](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/commit/cd2d5b988544246d87abbc441895d76003fed72b) to update create app templates.

### Patch Changes

- bbceed8b: Updates dependencies: @types/fs-extra, typescript, husky and @types/node
- Updated dependencies [13f39ac7]
- Updated dependencies [91ed3d75]
- Updated dependencies [72fc2752]
- Updated dependencies [e76d5fdd]
- Updated dependencies [bbceed8b]
  - @cloud-carbon-footprint/common@1.0.0

## 0.8.2

### Patch Changes

- e93d31ec: Extracts common data from cloud providers
- 3b42775b: Updates @types/node to 15.12.4
- Updated dependencies [e93d31ec]
- Updated dependencies [3b42775b]
- Updated dependencies [53366130]
  - @cloud-carbon-footprint/common@0.0.3

## 0.8.1

### Patch Changes

- 3f283737: Adds support for regions with unknown constants
- Updated dependencies [8e24c32b]
  - @cloud-carbon-footprint/common@0.0.2

## 0.8.0

### Minor Changes

- f9fbcb4c: Updates GCPCredentials class to us the @google-cloud/iam-credentials library instead, to reduce the core package size
- ababb826: Extracts two new packages app and common to avoid circular dependancies and make it easier to extract cloud provider packages
- e84a4c7a: Extract logic into the new packages: app, common, gcp, aws, azure:

  There are many files that have been updated/extracted.
  In order to update create-app templates, refer to the follow [commit](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/commit/8c6aaed52e9f3949e134852986d50362aad3367a).

  The following changes were made to,
  'packages/create-app/templates/default-app/packages/client/tsconfig.json':

  ```diff
      // ...
          "skipLibCheck": true,
          "esModuleInterop": true,
          "allowSyntheticDefaultImports": true,
  -        "strict": true,
          "forceConsistentCasingInFileNames": true,
          "module": "esnext",
          "moduleResolution": "node",
          "resolveJsonModule": true,
  -        "isolatedModules": true,
          "noEmit": true,
          "jsx": "react-jsx",
  -        "noFallthroughCasesInSwitch": true
  +        "noFallthroughCasesInSwitch": true,
  +        "strict": false,
  +        "isolatedModules": true
        },
  -      "include": [
  -        "src",
  -        "node_modules/apexcharts/types/apexcharts.d.ts"
  -      ]
  +      "include": ["src"]
      }
      // ...
  ```

  Additionally, the following dependencies have been updated and should also be updated in their respective template package.json file:

  - @cloud-carbon-footprint root package.json:
    - "@types/fs-extra": "^9.0.11"
    - "concurrently": "^6.2.0"
    - "marked": ">=2.0.5"
  - @cloud-carbon-footprint/api and @cloud-carbon-footprint/cli:
    - "dotenv": "^10.0.0"
    - "@cloud-carbon-footprint/app": Can be added with `yarn up @cloud-carbon-footprint/app`
    - "@cloud-carbon-footprint/common": Can be added with `yarn up @cloud-carbon-footprint/common`
  - @cloud-carbon-footprint/client:
    - "dotenv": "^10.0.0"
    - "@testing-library/react-hooks": "^7.0.0"
    - "concurrently": "^6.2.0"
    - "@cloud-carbon-footprint/common": Can be added with `yarn up @cloud-carbon-footprint/common`

### Patch Changes

- c7fa7db0: Updates dependencies to the latest

## 0.7.0

### Minor Changes

- 68365cbf: implements memory calculations for gcp
- 8df5703b: implements memory calculations for azure
- 3e2f876d: adds improvements to memory implementation for aws
- 370c509d: implements memory calculations for aws

### Patch Changes

- 56bb6da6: #244 Bug fixed by changing configs boolean conditionals:
  In order to update, you need to apply the following changes in the `client` package:

  `packages/client/src/dashboard/CarbonFormulaDrawer.tsx`:

  ```diff
      // ...

        }
      }

  -    const previousYearOfUsage = !!process.env.REACT_APP_PREVIOUS_YEAR_OF_USAGE
  +    const previousYearOfUsage =
  +        !!process.env.REACT_APP_PREVIOUS_YEAR_OF_USAGE &&
  +        process.env.REACT_APP_PREVIOUS_YEAR_OF_USAGE !== 'false'

      const appConfig: ClientConfig = {
          CURRENT_PROVIDERS: [

      // ...
  ```

- b63d8a67: The default `aws-sdk` dependency was bumped to `"^2.890.0"`,
- 3abe3dca: Adds support for AmazonNeptune storage service to AWS CostAndUSageReports
- 29f48e7c: Bug Fixed: Cache file is written with empty API data
- 7d523b59: adds default unknown processor types for azure and aws

## 0.6.1

### Patch Changes

- 021c345f: Google Cloud Storage option added to store cache file

## 0.6.0

### Minor Changes

- 8b4c992d: Add support for using the median for a group of processors when there are outliers

### Patch Changes

- 216ea2ec: adds error handling for azure request error"

## 0.5.0

### Minor Changes

- 0ccce96b: Adds additional Azure usage types and regions based on THoughtWorks' Azure usage

### Patch Changes

- 8175b41d: updates packages with readmes and metadata

## 0.4.0

### Minor Changes

- 6c620db: Updates the GCP Emissions factors to use those published by Google
- 033a504: adds integration testing
- 82aeb1e: Updates average min/max compute coefficients for cloud providers based on microarchitectures. Adds min/max compute coefficient for Graviton 2 processor

### Patch Changes

- f3d4c8a: fixes azure pue value
- c5f28fe: updates googleapis dependency version
- 906f14e: updates consumption types for azure and adds logs
- bb82d7b: updates gcp shared core processors list to enum
- 9b10f3b: fixes hdd storage types list and lint fix

## 0.3.0

### Minor Changes

- 29e9897: implements using min and max coefficients based on processor types for gcp
- 75176d7: Updates the cloud provider queries to be configurable by grouping dates and allows configurable date ranges
- ce0f249: Updates Application Config to support Docker Secrets

### Patch Changes

- beabadf: fix storage estimator and estimates cache
- e2ac070: Fixes default config for querying data
- 29e9897: adds AmazonLightsail to aws list of ssd services types
- a0ac965: Fixes failing tests when running them in a different timezone

## 0.2.0

### Minor Changes

- 96cb023: Adds support for Microsoft Azure, as well as updates to the microsite.

## 0.1.0

### Minor Changes

- 3b3bec3: Update packages to have base version
