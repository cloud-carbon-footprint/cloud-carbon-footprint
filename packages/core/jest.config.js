/*
 * Â© 2021 Thoughtworks, Inc.
 */

// eslint-disable-next-line @typescript-eslint/no-var-requires
const baseConfig = require('../../jest.base.config')

module.exports = {
  ...baseConfig,
  coverageThreshold: {
    global: {
      statements: 99,
      branches: 88,
      functions: 100,
      lines: 99,
    },
  },
  testPathIgnorePatterns: [
    '<rootDir>/src/__tests__/__mocks__',
    '<rootDir>/src/__tests__/fixtures',
  ],
  coveragePathIgnorePatterns: [
    '<rootDir>/src/index.ts',
    '<rootDir>/src/BillingDataRow.ts',
    '<rootDir>/src/FootprintEstimatesDataBuilder.ts',
    '<rootDir>/src/compute/ServiceWithCPUUtilization.ts',
    '<rootDir>/src/.*/index.ts',
  ],
}
