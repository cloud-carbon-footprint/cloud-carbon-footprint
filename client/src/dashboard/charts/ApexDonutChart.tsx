/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

import React, { FunctionComponent } from 'react'
import { useTheme } from '@material-ui/core/styles'
import Chart from 'react-apexcharts'

import { getChartColors } from '../../themes'
import { sumCO2ByServiceOrRegion } from '../transformData'
import { EstimationResult } from '../../models/types'

export const ApexDonutChart: FunctionComponent<ApexDonutChartProps> = ({ data, dataType }) => {
  const theme = useTheme()
  const chartColors = getChartColors(theme)

  const donutData = sumCO2ByServiceOrRegion(data, dataType)

  const options = {
    chart: {
      background: theme.palette.background.paper,
    },
    colors: chartColors,
    dataLabels: {
      style: {
        fontSize: '1rem',
      },
      dropShadow: {
        enabled: true,
        top: 0,
        left: 0,
        blur: 3,
        color: theme.palette.grey[800],
        opacity: 0.5,
      },
    },
    labels: Object.keys(donutData),
    legend: {
      position: 'bottom',
      offsetY: -8,
    },
    series: Object.values(donutData),
    stroke: {
      colors: [theme.palette.background.default],
    },
    title: {
      offsetY: -8,
      style: {
        fontSize: '24px',
      },
    },
    height: '500px',
    theme: {
      mode: theme.palette.type,
    },
    tooltip: {
      fillSeriesColor: false,
      y: {
        formatter: function (value: number) {
          return `${value.toFixed(3)} mt CO2e`
        },
      },
    },
  }

  return <Chart options={options} series={options.series} type="donut" height={options.height} />
}

type ApexDonutChartProps = {
  data: EstimationResult[]
  dataType: string
}
