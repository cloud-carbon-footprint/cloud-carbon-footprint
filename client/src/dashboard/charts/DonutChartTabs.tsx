import React from 'react'
import { Paper, Tabs, Tab, Box, Card } from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import { ApexDonutChart } from './ApexDonutChart'

const useStyles = makeStyles({
  root: {
    flexGrow: 1,
  },
  containerMargin: {
    bottom: 100,
  },
})

export const DonutChartTabs = (props: { data: any }) => {
  const classes = useStyles()
  const [value, setValue] = React.useState(0)

  const handleChange = (event: React.ChangeEvent<{}>, newValue: number) => {
    setValue(newValue)
  }

  return (
    <Card style={{ width: '100%', height: '100%' }}>
      <Box padding={3}>
        <Paper>
          <Tabs value={value} onChange={handleChange} indicatorColor="primary" textColor="primary" centered>
            <Tab id="Region" label="Emissions By Region" />
            <Tab id="Service" label="Emissions By Service" />
          </Tabs>
        </Paper>
        {value ? (
          <ApexDonutChart data={props.data} dataType="service" data-testid="service" />
        ) : (
          <ApexDonutChart data={props.data} dataType="region" data-testid="region" />
        )}
      </Box>
    </Card>
  )
}
