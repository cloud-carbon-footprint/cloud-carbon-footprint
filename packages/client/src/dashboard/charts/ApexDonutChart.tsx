/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

import React, { FunctionComponent } from 'react'
import { useTheme } from '@material-ui/core/styles'
import Chart from 'react-apexcharts'

import { getChartColors } from '../../themes'
import { sumCO2ByServiceOrRegion } from '../transformData'
import { ApexChartProps } from './common/ChartTypes'

export const ApexDonutChart: FunctionComponent<ApexChartProps> = ({ data, dataType }) => {
  const theme = useTheme()
  const chartColors = getChartColors(theme)

  const donutData = sumCO2ByServiceOrRegion(data, dataType)

  const donutDataEntries: { serviceOrRegion: string; c02Value: number }[] = Object.entries(donutData)
    .map((item) => ({
      serviceOrRegion: item[0],
      c02Value: item[1],
    }))
    .sort((higherC02, lowerCO2) => lowerCO2.c02Value - higherC02.c02Value)

  const sortedServicesOrRegions = donutDataEntries.map((entry) => entry.serviceOrRegion)
  const sortedCO2Emissions = donutDataEntries.map((entry) => entry.c02Value)

  const options = {
    chart: {
      background: theme.palette.background.paper,
    },
    colors: chartColors,
    dataLabels: {
      style: {
        fontSize: '1rem',
      },
      dropShadow: {
        enabled: true,
        top: 0,
        left: 0,
        blur: 3,
        color: theme.palette.grey[800],
        opacity: 0.5,
      },
    },
    labels: sortedServicesOrRegions,
    legend: {
      position: 'bottom',
      offsetY: -8,
    },
    series: sortedCO2Emissions,
    stroke: {
      colors: [theme.palette.background.default],
    },
    title: {
      offsetY: -8,
      style: {
        fontSize: '24px',
      },
    },
    height: '500px',
    theme: {
      mode: theme.palette.type,
    },
    tooltip: {
      fillSeriesColor: false,
      y: {
        formatter: function (value: number) {
          return `${value.toFixed(3)} metric tons CO2e`
        },
      },
    },
  }

  return <Chart options={options} series={options.series} type="donut" height={options.height} />
}
