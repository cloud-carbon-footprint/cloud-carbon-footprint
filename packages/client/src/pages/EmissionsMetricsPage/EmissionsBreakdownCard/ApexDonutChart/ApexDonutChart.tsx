/*
 * Â© 2021 Thoughtworks, Inc.
 */

import React, { FunctionComponent } from 'react'
import { useTheme } from '@material-ui/core/styles'
import Chart from 'react-apexcharts'

import { CCFTheme, getChartColors } from '../../../../utils/themes'
import { sumCO2ByServiceOrRegion } from '../../../../utils/helpers'
import { ApexChartProps } from '../../../../Types'

export const ApexDonutChart: FunctionComponent<ApexChartProps> = ({
  data,
  dataType,
}) => {
  const theme = useTheme() as CCFTheme
  const chartColors = getChartColors(theme)

  const donutData = sumCO2ByServiceOrRegion(data, dataType)

  const donutDataEntries: {
    serviceOrRegion: string
    c02Value: number
  }[] = Object.entries(donutData)
    .map((item) => ({
      serviceOrRegion: item[0],
      c02Value: item[1][1],
    }))
    .sort((higherC02, lowerCO2) => lowerCO2.c02Value - higherC02.c02Value)

  const sortedServicesOrRegions = donutDataEntries.map(
    (entry) => entry.serviceOrRegion,
  )
  const sortedCO2Emissions = donutDataEntries.map((entry) => entry.c02Value)

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const options: any = {
    chart: {
      background: theme.palette.background.paper,
    },
    colors: chartColors,
    dataLabels: {
      style: {
        fontSize: '1rem',
      },
      dropShadow: {
        enabled: true,
        top: 0,
        left: 0,
        blur: 3,
        color: theme.palette.grey[800],
        opacity: 0.5,
      },
    },
    labels: sortedServicesOrRegions,
    legend: {
      position: 'bottom',
      offsetY: -8,
    },
    series: sortedCO2Emissions,
    stroke: {
      colors: [theme.palette.background.default],
    },
    title: {
      offsetY: -8,
      style: {
        fontSize: '24px',
      },
    },
    height: '500px',
    theme: {
      mode: theme.palette.type,
    },
    tooltip: {
      fillSeriesColor: false,
      y: {
        formatter: function (value: number) {
          return `${value.toFixed(3)} metric tons`
        },
      },
    },
  }

  return (
    <Chart
      options={options}
      series={options.series}
      type="donut"
      height={options.height}
    />
  )
}
