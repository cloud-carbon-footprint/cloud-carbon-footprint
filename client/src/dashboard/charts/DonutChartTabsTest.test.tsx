import React from 'react'
import { create } from 'react-test-renderer'

import { DonutChartTabs } from './DonutChartTabs'
import { Tab } from '@material-ui/core'
import { ApexDonutChart } from './ApexDonutChart'
import Enzyme, { mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

jest.mock('../../themes')

describe('DonutChartTabs', () => {
  const date1 = new Date('2020-07-10T00:00:00.000Z')
  const date2 = new Date('2020-07-11T00:00:00.000Z')

  const dataWithHigherPrecision = [
    {
      timestamp: date1,
      serviceEstimates: [
        {
          timestamp: date1,
          serviceName: 'ebs',
          wattHours: 12.2342,
          co2e: 15.12341,
          cost: 5.82572,
          region: 'us-east-1',
          usesAverageCPUConstant: false,
        },
        {
          timestamp: date1,
          serviceName: 'ec2',
          wattHours: 4.745634,
          co2e: 5.234236,
          cost: 4.732,
          region: 'us-east-1',
          usesAverageCPUConstant: false,
        },
      ],
    },
    {
      timestamp: date2,
      serviceEstimates: [
        {
          timestamp: date2,
          serviceName: 'ebs',
          wattHours: 25.73446,
          co2e: 3.2600234,
          cost: 6.05931,
          region: 'us-east-1',
          usesAverageCPUConstant: false,
        },
        {
          timestamp: date2,
          serviceName: 'ec2',
          wattHours: 2.4523452,
          co2e: 7.7536,
          cost: 6.2323,
          region: 'us-east-1',
          usesAverageCPUConstant: true,
        },
      ],
    },
  ]

  it('renders donut chart with two tabs', () => {
    const testRenderer = create(<DonutChartTabs data={dataWithHigherPrecision} />)
    const testInstance = testRenderer.root
    const allTabInstancesList = testInstance.findAllByType(Tab)

    expect(allTabInstancesList).toHaveLength(2)

    allTabInstancesList.forEach((tab) => {
      expect(['Emissions By Region', 'Emissions By Service'].includes(tab.props.label)).toBe(true)
    })

    expect(testInstance.findAllByType(Tab))

    expect(testRenderer.toJSON()).toMatchSnapshot()
  })

  it('checks to see if donut chart exists upon loading', () => {
    const testRenderer = create(<DonutChartTabs data={dataWithHigherPrecision} />)
    const testInstance = testRenderer.root
    const isApexDonutChartRendered = testInstance.findAllByType(ApexDonutChart)

    expect(isApexDonutChartRendered).toHaveLength(1)
  })

  it('renders emission by region donut chart by default', () => {
    const testRenderer = create(<DonutChartTabs data={dataWithHigherPrecision} />)
    const testInstance = testRenderer.root
    const isApexDonutChartRendered = testInstance.findByType(ApexDonutChart)

    expect(isApexDonutChartRendered.props.dataType).toBe('region')
  })
  it('renders emission by service donut chart when service tab clicked', () => {
    Enzyme.configure({ adapter: new Adapter() })
    const wrapper = mount(<DonutChartTabs data={dataWithHigherPrecision} />)
    const tabsInstance = wrapper.find(Tab).at(1).simulate('click')
    wrapper.update()

    const isApexDonutChartRendered = wrapper.find(ApexDonutChart)
    expect(isApexDonutChartRendered.props().dataType).toBe('service')
  })
})
