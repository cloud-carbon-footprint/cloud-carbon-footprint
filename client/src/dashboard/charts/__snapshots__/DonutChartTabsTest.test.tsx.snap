// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DonutChartTabs renders donut chart with two tabs 1`] = `
<div
  className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
>
  <div
    className="MuiBox-root MuiBox-root-3"
  >
    <div
      className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        className="MuiTabs-root"
      >
        <div
          className="MuiTabs-scroller MuiTabs-fixed"
          onScroll={[Function]}
          style={
            Object {
              "marginBottom": null,
              "overflow": "hidden",
            }
          }
        >
          <div
            className="MuiTabs-flexContainer MuiTabs-centered"
            onKeyDown={[Function]}
            role="tablist"
          >
            <button
              aria-selected={true}
              className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected"
              disabled={false}
              id="Region"
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              role="tab"
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiTab-wrapper"
              >
                Emissions By Region
              </span>
              <span
                className="PrivateTabIndicator-root-4 PrivateTabIndicator-colorPrimary-5 MuiTabs-indicator"
                style={Object {}}
              />
            </button>
            <button
              aria-selected={false}
              className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary"
              disabled={false}
              id="Service"
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              role="tab"
              tabIndex={-1}
              type="button"
            >
              <span
                className="MuiTab-wrapper"
              >
                Emissions By Service
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      data-testid="fake-donut-chart"
      height="500px"
      options={
        Object {
          "chart": Object {
            "background": "#fff",
          },
          "colors": Array [
            "#000000",
            "#FFFFFF",
          ],
          "dataLabels": Object {
            "dropShadow": Object {
              "blur": 3,
              "color": "#424242",
              "enabled": true,
              "left": 0,
              "opacity": 0.5,
              "top": 0,
            },
            "style": Object {
              "fontSize": "1rem",
            },
          },
          "height": "500px",
          "labels": Array [
            "us-east-1",
          ],
          "legend": Object {
            "offsetY": -8,
            "position": "bottom",
          },
          "series": Array [
            31.3712694,
          ],
          "stroke": Object {
            "colors": Array [
              "#fafafa",
            ],
          },
          "theme": Object {
            "mode": "light",
          },
          "title": Object {
            "offsetY": -8,
            "style": Object {
              "fontSize": "24px",
            },
          },
          "tooltip": Object {
            "fillSeriesColor": false,
            "y": Object {
              "formatter": [Function],
            },
          },
        }
      }
      series={
        Array [
          31.3712694,
        ]
      }
      type="donut"
    >
      fake chart
    </div>
  </div>
</div>
`;
